### ISSUE TITLE
Compilation cancelation

### ISSUE THREAD

**Issue by rauchg** [vercel/next.js#1083]
Sometimes you make a change, then want to quickly make another change
Webpack doesn't seem to do cancelation by default? Which means a lot of wasted time!

**Comment by rauchg** [comment]
+1. This is really really annoying

**Comment by rauchg** [comment]
@msand thanks for pinging that issue. 

I just realized not having cancelation is not only a "nice to have", but it's absolutely critical for scenarios like changes that happen in quick succession. For example, `git checkout`.

**Comment by componentDidMount** [comment]
Maybe during the build something could listen (somehow connected to watching file saves/changes) and throw an error, so the build fails/stops, so before every save of a file the current build listening (if there is one running) would fail before the next one starts. Would that somehow be an OK solution to this issue?

**Comment by quietshu** [comment]
Just saw https://github.com/webpack/webpack/issues/1702 this issue and @TheLarkInn mentioned that Nuxt.js does great for this. 

So I read its code and seems Nuxt.js is using chokidar as the file watcher rather than Webpack, and also it's using `_.debounce` to avoid rebuild caused by quick file changes (see https://github.com/nuxt/nuxt.js/blob/master/lib/builder/builder.js#L660). But it also increases the delay by a little bit.

Webpack supports this "rebuild delay" and we can add this feature to Next.js really quickly by add `watchOptions.aggregateTimeout` (see https://webpack.js.org/configuration/watch/#watchoptions-aggregatetimeout) here: https://github.com/zeit/next.js/blob/canary/server/hot-reloader.js#L223.

What do you say? I think this might be very useful.

**Comment by quietshu** [comment]
I'd like to try implementing a demo using the idea of Webpack's `watching.invalidate()` and chokida.

**Comment by quietshu** [comment]
Hmmm seems really hard to implement it. I digged into the Webpack code and it turns out that, in the watch mode, although there's an undelayed callback for the file watcher, but the file change event will still be blocked by the current compilation. Maybe it's because the ongoing compilation and the file watcher run in the same single thread.

Same issue here: https://github.com/webpack/webpack/issues/5929
